### ISSUE TITLE
Highlighting "invalid" code via semantic tokens (vs diagnostics)

### ISSUE THREAD

**Issue by radeksimko** [microsoft/language-server-protocol#1415]
The protocol specification doesn't seem to define use cases for semantic token based highlighting too explicitly. The only part of spec documenting use case is AFAIK the following sentence.

> Semantic tokens are used to add additional color information to a file that depends on language specific symbol information.

That may define the relationship between static grammar such as TextMate in VSCode and LSP, where LSP is assumed to **enrich**, rather than **replace**. Although even that probably can't be taken for granted depending on `augmentsSyntaxTokens` client capability, which could result in semantic tokens being the only source of highlighting. In that case the colours are no longer "additional".

When it comes to highlighting some basic invalid code in TextMate, it documents the following relevant scopes:

> - `invalid` — stuff which is “invalid”.
>   - `illegal` — illegal, e.g. an ampersand or lower-than character in HTML (which is not part of an entity/tag).
>   - `deprecated` — for deprecated stuff e.g. using an API function which is deprecated or using styling with strict HTML.

https://macromates.com/manual/en/language_grammars

Default VSCode themes only seem to reflect `invalid`, which ends up as red colour.

The lack of relevant token types/modifiers in LSP here would imply that semantic tokens are not really used for reporting invalid code. However was this a conscious design decision, or is this something that could be added later e.g. as part of https://github.com/microsoft/language-server-protocol/issues/1052 or https://github.com/microsoft/language-server-protocol/issues/1067 ?

The bigger question is "What's even _invalid_?" One could consider deprecated code as invalid too - which LSP already covers via [`deprecated` modifier](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#semanticTokenModifiers).

Invalid code seems (from my own experience) generally more often dealt with by diagnostics, which clients would typically highlight differently anyway (as red & squiggly underline). Relatedly LSP 3.15.0 also introduced [`Deprecated` DiagnosticTag](https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#diagnosticTag)). This is further supported by the fact that diagnostics can provide context on _why_ something is invalid, rather than just giving piece of code different style, so it would seem to me this is the preferred way.

TL;DR Where does one draw the line between invalid code reported via **diagnostics** VS **semantic tokens**?

**Comment by rcjsuen** [comment]
> TL;DR Where does one draw the line between invalid code reported via **diagnostics** VS **semantic tokens**?

I think I would draw the line at whether I want this issue reported to the user in another place (such as a "Problems" panel/view/section/whatever) in the client or if it should only ever be visible inside an editor.

When the editor is opened on file `a.txt` but `b.txt` has an error should the user be aware of this error? If yes, then it should be a diagnostic. Semantic tokens will not help them in this case because `b.txt` has not been opened.

**Comment by puremourning** [comment]
I don't think semantic tokens should ever be used to indicate errors or problems. That's what diagnostics are for. 

semantic tokens are complex and _optional_ feature that many clients don't or won't im0plement. diagnostics are a primary mandatory feature.

**Comment by dbaeumer** [comment]
I totally agree. I will close the issue since there are no plans to work on this.

**Comment by radeksimko** [comment]
That is fair 👍🏻  but I'd say that it already encourages highlighting of invalid tokens through the mentioned `deprecated` modifier, so perhaps the spec needs a comment to say that diagnostics should be preferred, and/or even deprecate the deprecated modifier in favour of diagnostics? I'd be happy to propose that change in a PR.

**Comment by dbaeumer** [comment]
I am always happy to accept clarification PRs
